File: plugin/src/main/java/com/jtprince/bingo/plugin/Space.java
package com.jtprince.bingo.plugin;

import com.jtprince.bingo.plugin.automarking.AutoMarkTrigger;
import org.json.simple.JSONObject;

import java.util.Collection;
import java.util.HashMap;
import java.util.Map;

public class Space {
    public final GameBoard board;

    public final String goalId;
    public final String goalType;
    public final String text;
    public final int spaceId;
    public final Map<String, Integer> variables = new HashMap<>();

    public final Collection<AutoMarkTrigger> autoMarkTriggers;

    public Space(GameBoard board, String goalId, String goalType, String text, int spaceId,
                 Map<String, Integer> variables) {
        this.board = board;
        this.goalId = goalId;
        this.goalType = goalType;
        this.text = text;
        this.spaceId = spaceId;
        if (variables != null) {
            this.variables.putAll(variables);
        }

        this.autoMarkTriggers = AutoMarkTrigger.createAllTriggers(this);
    }

    public Space(GameBoard board, JSONObject obj) {
        this(board, (String) obj.get("goal_id"), (String) obj.get("type"),
            (String) obj.get("text"), ((Long) obj.get("space_id")).intValue(),
            parseVariablesMap((JSONObject) obj.get("variables")));
    }

    public void destroy() {
        for (AutoMarkTrigger t : this.autoMarkTriggers) {
            t.destroy();
        }
    }

    public boolean isAutoMarked() {
        return autoMarkTriggers.size() > 0;
    }

    private static Map<String, Integer> parseVariablesMap(JSONObject variables) {
        Map<String, Integer> ret = new HashMap<>();
        if (variables != null) {
            for (Object o : variables.keySet()) {
                String k = (String) o;
                int v = ((Long) variables.get(k)).intValue();
                ret.put(k, v);
            }
        }
        return ret;
    }
}

File: kotlinplugin/src/main/kotlin/com/jtprince/bingo/kplugin/board/Space.kt
package com.jtprince.bingo.kplugin.board

import com.jtprince.bingo.kplugin.automark.AutoMarkTrigger
import com.jtprince.bingo.kplugin.automark.AutoMarkTriggerFactory
import com.jtprince.bingo.kplugin.automark.AutomatedSpace
import com.jtprince.bingo.kplugin.automark.PlayerTriggerProgress
import com.jtprince.bingo.kplugin.game.PlayerManager
import com.jtprince.bingo.kplugin.player.BingoPlayer

class Space(
    override val spaceId: Int,
    override val goalId: String,
    val goalType: GoalType,
    override val text: String,
    override val variables: SetVariables,
) : AutomatedSpace {
    enum class Marking(val value: Int) {
        UNMARKED(0),
        COMPLETE(1),
        REVERTED(2),
        INVALIDATED(3),
        NOT_INVALIDATED(4);

        companion object {
            private val map = values().associateBy(Marking::value)
            fun valueOf(value: Int) = map[value] ?:
                throw IllegalArgumentException("Unknown Marking value $value")
        }
    }

    enum class GoalType(val value: String) {
        DEFAULT("default"),
        NEGATIVE("negative");

        companion object {
            private val map = values().associateBy(GoalType::value)
            fun ofString(value: String) = map[value] ?:
                throw IllegalArgumentException("Unknown Marking value $value")
        }
    }

    private lateinit var triggers: Collection<AutoMarkTrigger>

    val automarking: Boolean
        get() = triggers.isNotEmpty()

    /**
     * Receive callbacks when a player in this [PlayerManager] does something that should mark the
     * space.
     */
    fun startListening(playerManager: PlayerManager, callback: AutoMarkTrigger.Callback) {
        triggers = AutoMarkTriggerFactory().create(this, playerManager, callback)
    }

    /**
     * Stop receiving callbacks for all players on this space.
     */
    fun stopListening() {
        triggers.forEach(AutoMarkTrigger::destroy)
    }

    fun destroy() {
        stopListening()
    }

    override val playerProgress: MutableMap<BingoPlayer, PlayerTriggerProgress> by lazy { mutableMapOf() }
}
